### DISCUSSION TITLE
Iterator like execution with Dynamic Task Mapping?

### DISCUSSION THREAD

**Discussion by zegab1** [apache/airflow#28162] [Ideas]
I have a use case where I need to process data sets in sequential order, as the processing a dataset depends on the results from the previous dataset. I love the simplicity and elegance of the new Dynamic Task Mapping feature, but unfortunately I cannot use it here. 

For now, I have a single Spark job wich takes care of iterating through all the requred datasets, but for the use case, it would be much better if there would be a separate spark job for processing each dataset.

I wonder if we could have this as a feature, either as a new function (like iterate()), or as an option to expand() that would cause all mapped tasks to be executed sequentially?

**Comment by potiuk** [comment]
Create a pool of size 1 and mark your dynamic tasks to use the pool

**Reply by zegab1** [reply]
Is it guarantees in-order execution of the dynamic tasks in all cases?

**Reply by potiuk** [reply]
No. If you want to guarantee the order of execution you should use regular Dynamic Dags and create your tasks in a for loop and add dependencies. This is basic feature of airflow.

**Reply by zegab1** [reply]
I need to figure out the list of dataset IDs at runtime, and I need to iterate over this list. If I had no dependency on execution order, Dynamic Task Mapping could handle that well, as I could have a task to get the list, and fed it to expand() to generate the tasks.

With standard dynamically generated DAGs, this would be more complex, I would need to do some tricks to achieve the same functionality. Like to have a loop N times (where N is the max number of expected iterations), in each loop, I need to figure out if this is a valid loop or not, then get the dataset ID for that iteration, execute the processing task. Or maybe just have a DAG to get the dataset IDs, dynamically generate a new python file with a new DAG that would process these specific IDs, trigger the 2nd DAG from the first, then remove it. 

I think both solutions are complex, hard to understand and maintain. That's why I have this iteration right now in the Spark code, and not in Airflow. The drawback of this approach is that this is a single Spark job, it can take a long time to execute, and if it fails, the whole task will be executed again. Braking it down to smaller tasks in Airflow would make it more resilient, but right now the added complexity of the implementation is counterbalancing the gains for me.

So I still thing having an "iterate()" function that could process result from an upstream task is a reasonable feature.

**Reply by potiuk** [reply]
I personally think it's a bit of a niche case, but ultimately - you might be able to convince others in a community if you make the effort, and it's not a case that I would veto if somoeone implements it.

This is an open-source project with huge community and clear governance rules. Such a change would likely require to make an AIP proposal https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals for it and discuss the details on the devlist. I think such a feature requires quite a bit of a thinking how to implement it, discussing consequences and proposing how the interface would look like. 

That requires justifying and explaining the "general" cases that you want to handle it wiht this proposal. It would also mean that you will need to gather support for your idea and eventually call for a vote and then it would have to get implemented (either by you or by people you will find supporting your idea and wanting to implement it).
